source/extensions/access_loggers/common/grpc_access_logger_clients.h

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/config/core/v3/config_source.pb.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/grpc/async_client_manager.h"
+#include "envoy/singleton/instance.h"
+#include "envoy/stats/scope.h"
+#include "envoy/thread_local/thread_local.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/grpc/typed_async_client.h"
+#include "source/common/http/utility.h"
+#include "source/common/protobuf/utility.h"
+#include "source/common/tracing/null_span_impl.h"
+#include "source/extensions/access_loggers/common/grpc_access_logger_utils.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace AccessLoggers {
+namespace Common {
+
+template <typename LogRequest, typename LogResponse> class GrpcAccessLogClient {
+public:
+  virtual ~GrpcAccessLogClient() = default;
+  virtual bool isConnected() PURE;
+  virtual bool log(const LogRequest& request) PURE;
+
+protected:
+  GrpcAccessLogClient(const Grpc::RawAsyncClientSharedPtr& client,
+                      const Protobuf::MethodDescriptor& service_method,
+                      OptRef<const envoy::config::core::v3::RetryPolicy> retry_policy)
+      : client_(client), service_method_(service_method),
+        opts_(createRequestOptionsForRetry(retry_policy)) {}
+
+  Grpc::AsyncClient<LogRequest, LogResponse> client_;
+  const Protobuf::MethodDescriptor& service_method_;
+  const Http::AsyncClient::RequestOptions opts_;
+
+private:
+  Http::AsyncClient::RequestOptions
+  createRequestOptionsForRetry(OptRef<const envoy::config::core::v3::RetryPolicy> retry_policy) {
+    auto opt = Http::AsyncClient::RequestOptions();
+
+    if (!retry_policy) {
+      return opt;
+    }
+
+    const auto grpc_retry_policy =
+        Http::Utility::convertCoreToRouteRetryPolicy(*retry_policy, "connect-failure");
+    opt.setBufferBodyForRetry(true);
+    opt.setRetryPolicy(grpc_retry_policy);
+    return opt;
+  }
+};
+
+template <typename LogRequest, typename LogResponse>
+class UnaryGrpcAccessLogClient : public GrpcAccessLogClient<LogRequest, LogResponse> {
+public:
+  typedef std::function<Grpc::AsyncRequestCallbacks<LogResponse>*()> AsyncRequestCallbacksFactory;
```

Comment:
I think now you can make the factory return std::unique_ptr. This will indicate that there is an owner for the object.

Reply:
We are letting OpenTelemetry::GrpcAccessLoggerImpl to owns the callback so we only need to return pointer/reference?

Reply:
Ok, I see how you did it. I would just return reference instead of pointer then to signify that the callback object is owned elsewhere.

Reply:
DONE.